In [1]:
#2019243064 박규민 Assignment_Fashion_MNist
import tensorflow as tf
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras import datasets, layers, models
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D

fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

2023-05-30 22:49:43.539491: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.


In [2]:
train_images = train_images.reshape((60000, 28, 28, 1))
test_images = test_images.reshape((10000, 28, 28, 1))

train_images = train_images / 255
test_images = test_images / 255

model = models.Sequential()

model.add(layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))

model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))

#1 차원으로 전환 후 
model.add(layers.Flatten())

model.add(layers.Dense(128, activation='relu'))
#드롭아웃 기법으로 노이즈 제거
model.add(Dropout(0.2))

model.add(layers.Dense(10, activation='softmax'))

2023-05-30 22:49:47.392294: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-30 22:49:47.803164: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 9643 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:68:00.0, compute capability: 7.5


In [3]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 64)        640       
                                                                 
 conv2d_1 (Conv2D)           (None, 24, 24, 64)        36928     
                                                                 
 max_pooling2d (MaxPooling2D  (None, 12, 12, 64)       0         
 )                                                               
                                                                 
 conv2d_2 (Conv2D)           (None, 10, 10, 128)       73856     
                                                                 
 flatten (Flatten)           (None, 12800)             0         
                                                                 
 dense (Dense)               (None, 128)               1638528   
                                                        

In [4]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [5]:
model.fit(train_images, train_labels, batch_size=100, epochs=5)

Epoch 1/5


2023-05-30 22:49:52.872230: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8401


600/600 [==============================] - 5s 7ms/step - loss: 0.4488 - accuracy: 0.8381
Epoch 2/5
600/600 [==============================] - 4s 7ms/step - loss: 0.2677 - accuracy: 0.9036
Epoch 3/5
600/600 [==============================] - 4s 7ms/step - loss: 0.2205 - accuracy: 0.9188
Epoch 4/5
600/600 [==============================] - 4s 7ms/step - loss: 0.1860 - accuracy: 0.9309
Epoch 5/5
600/600 [==============================] - 4s 7ms/step - loss: 0.1574 - accuracy: 0.9414


In [7]:
model.fit(train_images, train_labels, batch_size=100, epochs=5)

Epoch 1/5
600/600 [==============================] - 4s 7ms/step - loss: 0.1287 - accuracy: 0.9519
Epoch 2/5
600/600 [==============================] - 4s 7ms/step - loss: 0.1105 - accuracy: 0.9587
Epoch 3/5
600/600 [==============================] - 4s 7ms/step - loss: 0.0900 - accuracy: 0.9669
Epoch 4/5
600/600 [==============================] - 4s 7ms/step - loss: 0.0774 - accuracy: 0.9710
Epoch 5/5
600/600 [==============================] - 4s 7ms/step - loss: 0.0663 - accuracy: 0.9749


In [9]:
model.fit(train_images, train_labels, batch_size=100, epochs=2)

Epoch 1/2
600/600 [==============================] - 4s 7ms/step - loss: 0.0559 - accuracy: 0.9788
Epoch 2/2
600/600 [==============================] - 4s 7ms/step - loss: 0.0492 - accuracy: 0.9815


In [11]:
model.fit(train_images, train_labels, batch_size=100, epochs=3)

Epoch 1/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0470 - accuracy: 0.9821
Epoch 2/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0410 - accuracy: 0.9840
Epoch 3/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0361 - accuracy: 0.9865


In [13]:
model.fit(train_images, train_labels, batch_size=100, epochs=3)

Epoch 1/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0359 - accuracy: 0.9864
Epoch 2/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0322 - accuracy: 0.9883
Epoch 3/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0309 - accuracy: 0.9887


In [17]:
model.fit(train_images, train_labels, batch_size=100, epochs=3)

Epoch 1/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0296 - accuracy: 0.9896
Epoch 2/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0251 - accuracy: 0.9907
Epoch 3/3
600/600 [==============================] - 4s 7ms/step - loss: 0.0248 - accuracy: 0.9911


In [18]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 [==============================] - 1s 5ms/step - loss: 0.4989 - accuracy: 0.9238


### 아무리 학습 및 dense 층을 변경하여도 0.92~3 이 최대라서 이렇게 제출하게 되었습니다
### 학습은 최대한 가능한데까지 해두었습니다.(Overfit 피하기 위함...)